### PR TITLE
Refactor and repackage client/driver

### PR DESCRIPTION
**Pull request by dantoml**
This PR migrates future relevant code out of the client/driver/* package, and then removes the legacy driver interfaces.

A side effect of this is that we no longer get submission time job validation, but that should be coming as a follow up to the 0.9 release with included support for driver plugin validation.

❓
- Can we rm the allocrunnerdeprecated (and executor_pluginpre_09_0) trees now?

### CODE REVIEW

File: drivers/shared/env/util_test.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 
 // TestaddNestedKey_Ok asserts test cases that succeed when passed to
```
Comment (schmichael) [inline drivers/shared/env/util_test.go]:
```suggestion
// TestAddNestedKey_Ok asserts test cases that succeed when passed to
```

File: drivers/shared/structs/structs.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package structs
+
+import "errors"
+
+// CheckBufSize is the size of the buffer that is used for job output
+const CheckBufSize = 4 * 1024
+
+// DriverStatsNotImplemented is the error to be returned if a driver doesn't
+// implement stats.
+var DriverStatsNotImplemented = errors.New("stats not implemented for driver")
```
Comment (schmichael) [inline drivers/shared/structs/structs.go]: Wow, are these really the only things in this package? Could they be moved to `client/structs` with causing import cycles?

File: vendor/github.com/hashicorp/serf/serf/keymanager.go [hunk 3]
```diff
@@ -189,4 +189,4 @@ func (k *KeyManager) ListKeysWithOptions(opts *KeyRequestOptions) (*KeyResponse,
 	defer k.l.RUnlock()
 
 	return k.handleKeyRequest("", listKeysQuery, opts)
-}
\ No newline at end of file
+}
```
Comment (schmichael) [inline vendor/github.com/hashicorp/serf/serf/keymanager.go]: Mind reverting these vendor changes? I think govendor hashes vendored packages, so even single byte changes might throw something off.

Comment (dantoml) [reply vendor/github.com/hashicorp/serf/serf/keymanager.go]: oops yeah that must've been an accidental overreaching gofmt

File: plugins/executor/plugins.go [hunk 4]
```diff
@@ -8,6 +8,20 @@ import (
...
+
+	// LogLevel is the level of the logs to putout
+	LogLevel string
+
+	// FSIsolation if set will use an executor implementation that support
+	// filesystem isolation
+	FSIsolation bool
+}
+
 var HandshakeConfig = plugin.HandshakeConfig{
```
Comment (nickethier) [inline plugins/executor/plugins.go]: While you're here lets remove this and use `base.Handshake`.

File: plugins/drivers/utils/utils.go [hunk 5]
```diff
@@ -79,8 +78,8 @@ func CreateExecutor(w io.Writer, level hclog.Level, driverConfig *base.ClientDri
 	config := &plugin.ClientConfig{
 		Cmd: exec.Command(bin, "executor", string(c)),
 	}
-	config.HandshakeConfig = driver.HandshakeConfig
-	config.Plugins = driver.GetPluginMap(w, level, executorConfig.FSIsolation)
+	config.HandshakeConfig = pexecutor.HandshakeConfig
```
Comment (nickethier) [inline plugins/drivers/utils/utils.go]: See comment on using `base.Handshake`

File: drivers/shared/structs/structs.go (hunk 2 above)
Comment (nickethier) [reply drivers/shared/structs/structs.go]: Keeping CheckBufSize in `plugins/drivers/driver.go` and moving this error there seems reasonable as well. That package is already sort of the shared driver structs package.

Comment (dantoml) [reply drivers/shared/structs/structs.go]: The problem with putting them in `plugins/drivers` is that they're required by the executor implementation so you end up with:

```
package github.com/hashicorp/nomad
        imports github.com/hashicorp/nomad/command
        imports github.com/hashicorp/nomad/client/allocrunner/taskrunner/restarts
        imports github.com/hashicorp/nomad/plugins/drivers
        imports github.com/hashicorp/nomad/plugins/drivers/utils
        imports github.com/hashicorp/nomad/drivers/shared/executor
        imports github.com/hashicorp/nomad/plugins/drivers
```

Comment (dantoml) [reply drivers/shared/structs/structs.go]: They merge cleanly into client/structs tho 👍

File: vendor/github.com/hashicorp/serf/serf/keymanager.go (hunk 3 above)
Comment (dantoml) [reply vendor/github.com/hashicorp/serf/serf/keymanager.go]: fixed

File: plugins/executor/plugins.go (hunk 4 above)
Comment (dantoml) [reply plugins/executor/plugins.go]: Done

**Comment by dantoml** [comment]
@nickethier I've merged driver/shared/structs into client/structs, but they can't go into plugins/driver due to being required by parts of drivers (specifically executors).

**Pull request merged**